### PR TITLE
Switch to State - Label dependency

### PR DESCRIPTION
**Pull request by ec**

### CODE REVIEW

**Comment by mod** [comment]
- [x] Remove discared and put locked
- [x] Label are not limited to :verified we can define scope=system
- [x] move data field and allow json and max 1MB
- [x] Publish data in entity, remove extended user profile, default api is always extended
- [x] locked example label suspicious:trade suspicious:deposit

**Comment by mod** [comment]
```
user_state:
  requirements:
    active:
      email: 'verified'
      phone: 'verified'
  mapping:
    locked:
       suspicious: 'trade'
       suspicious: 'deposit'
```

**Comment by mod** [comment]
```
user_state:
  active: 
    all:
      - email: 'verified'
      - phone: 'verified'
   locked:
     any:
       - suspicious: 'trade'
       - suspicious: 'deposit'
```

**Comment by ec** [comment]
I suggest to have a module structure:
   module all
      state
      state
  module any
      state
      state

Looks more readable in code and not so bad in yml
```
user_state:
  all:
    active:
      email: 'verified'
      phone: 'verified'
    banned:
      ban: 'weneedtochoose'
    deleted:
      deleted: 'weneedtochoose'
  any:
    locked:
      trade: 'suspicious'
      deposit: 'suspicious'
```

File: app/api/v2/admin/users.rb [hunk 1]
```diff
@@ -98,6 +98,11 @@ def search(field, value)
 
             error!({ errors: ['admin.user.update_himself'] }, 422) if target_user.uid == current_user.uid
 
+            if update_param_key == 'state' && update_param_value == 'banned'
+              target_user.add_private_label('ban', 'weneedtochoose')
```
Comment (mod) [inline app/api/v2/admin/users.rb]: I'd rather not chooose; why is this necessary ?

File: config/user_state_requirements.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      ban: 'weneedtochoose'
+    # labels required for user to become deleted
+    deleted:
+      deleted: 'weneedtochoose'
+  # requires 1 or more labels
+  any:
+    # labels required for user to become locked
+    locked:
+      suspicious: 'trade'
+      suspicious: 'deposit'
```
Comment (mod) [inline config/user_state_requirements.yml]: config/barong.yml
```
activation_requirements:
      email: 'verified'
      phone: 'verified'
state_triggers:
    banned:
      - ban
      - fraud
    deleted:
      - delete
    locked:
      - suspicious
      - lock
feature_xyz:
  settings:
     sdfkljdlkfgj: 24
```

later:
barong.state_triggers.banned = [ban, fraud]
app_name . component . key = value

fraud:trade
fraud:deposit
fraud:*

**Review by mod — CHANGES_REQUESTED** [review]

File: app/api/v2/resource/users.rb [hunk 3]
```diff
@@ -41,8 +41,8 @@ def verify_otp!
 
           verify_otp! if current_user.otp
 
-          current_user.update(state: 'discarded')
-          EventAPI.notify('system.user.account.discarded', current_user.as_json_for_event_api)
+          current_user.labels.create(key: 'deleted', value: 'by_user', scope: 'private')
```
Comment (mod) [inline app/api/v2/resource/users.rb]: delete:by_user

File: app/models/user.rb [hunk 4]
```diff
@@ -13,8 +13,10 @@ class User < ApplicationRecord
   has_many  :api_keys,   dependent: :destroy, class_name: 'APIKey'
   has_many  :activities, dependent: :destroy
 
+  validates_length_of :data, :maximum => 1024
   validate :role_exists
   validate :referral_exists
+  validates :data, data_prefix: true
```
Comment (mod) [inline app/models/user.rb]: remove data_prefix: true

File: app/models/user.rb [hunk 5]
```diff
@@ -88,9 +83,24 @@ def update_level
...
-          .update!(value: value)
+  def update_state
+    @resulting_state = 'pending'
+
+    unless active?
+      if BarongConfig.list['activation_requirements'] <= (labels.with_private_scope.map { |l| { l.key => l.value }}.inject(:merge) || {})
+        @resulting_state = 'active'
+      end
+    end
+
```
Comment (mod) [inline app/models/user.rb:94]: #FIXME BarongConfig should be a feature of Barong::App
A good name for Barong::App can be Barong::Config and we can add database support or vault based configs.

File: config/initializers/barong_config.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+class BarongConfig
```
Comment (mod) [inline config/initializers/barong_config.rb:2]: FIXME Move to barong.rb later

**Review by mod — CHANGES_REQUESTED** [review]
almost ready for merge

**Pull request merged**